File: swift/Sources/CantonKit/CommandSubmission.swift
import CantonLedgerAPI
import Foundation
import SwiftProtobuf

/// A batch of commands to submit atomically to the ledger.
///
/// ``commandId`` defaults to a fresh UUID and stays stable for the lifetime
/// of this instance — the SDK's retries reuse it, so the participant's
/// command deduplication prevents double execution. Reuse one
/// `CommandSubmission` per logical action; build a new one for a genuinely
/// new action.
public struct CommandSubmission: Sendable {
    /// The commands to execute atomically, in order.
    public var commands: [Com_Daml_Ledger_Api_V2_Command]

    /// Parties on whose behalf the commands are executed.
    public var actAs: [String]

    /// Additional parties whose contracts may be read during interpretation.
    public var readAs: [String]

    /// The ledger user submitting the request; must match the JWT's user on
    /// authenticated ledgers.
    public var userId: String

    /// Unique id for deduplication; keep stable across retries of the same action.
    public var commandId: String

    public var workflowId: String

    /// How far back the participant rejects duplicate ``commandId``s;
    /// participant maximum if nil.
    public var deduplicationDuration: Duration?

    /// Pin execution to a synchronizer; participant chooses if empty.
    public var synchronizerId: String

    public init(
        commands: [Com_Daml_Ledger_Api_V2_Command],
        actAs: [String],
        readAs: [String] = [],
        userId: String = "",
        commandId: String = UUID().uuidString,
        workflowId: String = "",
        deduplicationDuration: Duration? = nil,
        synchronizerId: String = ""
    ) {
        precondition(!commands.isEmpty, "commands must not be empty")
        precondition(!actAs.isEmpty, "actAs must contain at least one party")
        self.commands = commands
        self.actAs = actAs
        self.readAs = readAs
        self.userId = userId
        self.commandId = commandId
        self.workflowId = workflowId
        self.deduplicationDuration = deduplicationDuration
        self.synchronizerId = synchronizerId
    }

    var proto: Com_Daml_Ledger_Api_V2_Commands {
        var proto = Com_Daml_Ledger_Api_V2_Commands()
        proto.commandID = commandId
        proto.userID = userId
        proto.workflowID = workflowId
        proto.synchronizerID = synchronizerId
        proto.commands = commands
        proto.actAs = actAs
        proto.readAs = readAs
        if let dedup = deduplicationDuration {
            var duration = Google_Protobuf_Duration()
            duration.seconds = dedup.components.seconds
            duration.nanos = Int32(dedup.components.attoseconds / 1_000_000_000)
            proto.deduplicationDuration = duration
        }
        return proto
    }
}

File: kotlin/canton-sdk/src/main/kotlin/io/github/vsima/canton/CommandSubmission.kt
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

package io.github.vsima.canton

import com.daml.ledger.api.v2.CommandsOuterClass
import java.util.UUID
import kotlin.time.Duration
import com.google.protobuf.Duration as ProtoDuration

/**
 * A batch of commands to submit atomically to the ledger.
 *
 * [commandId] defaults to a fresh UUID and stays stable for the lifetime of
 * this instance — the SDK's retries reuse it, so the participant's command
 * deduplication prevents double execution. Reuse one [CommandSubmission] per
 * logical action; build a new one for a genuinely new action.
 */
public data class CommandSubmission(
    /** The commands to execute atomically, in order. */
    val commands: List<CommandsOuterClass.Command>,
    /** Parties on whose behalf the commands are executed. */
    val actAs: List<String>,
    /** Additional parties whose contracts may be read during interpretation. */
    val readAs: List<String> = emptyList(),
    /** The ledger user submitting the request; must match the JWT's user on authenticated ledgers. */
    val userId: String = "",
    /** Unique id for deduplication; keep stable across retries of the same action. */
    val commandId: String = UUID.randomUUID().toString(),
    val workflowId: String = "",
    /** How far back the participant rejects duplicate [commandId]s; participant maximum if null. */
    val deduplicationDuration: Duration? = null,
    /** Pin execution to a synchronizer; participant chooses if empty. */
    val synchronizerId: String = "",
) {
    init {
        require(commands.isNotEmpty()) { "commands must not be empty" }
        require(actAs.isNotEmpty()) { "actAs must contain at least one party" }
    }

    internal fun toProto(): CommandsOuterClass.Commands =
        CommandsOuterClass.Commands.newBuilder()
            .setCommandId(commandId)
            .setUserId(userId)
            .setWorkflowId(workflowId)
            .setSynchronizerId(synchronizerId)
            .addAllCommands(commands)
            .addAllActAs(actAs)
            .addAllReadAs(readAs)
            .apply {
                this@CommandSubmission.deduplicationDuration?.let { dedup ->
                    setDeduplicationDuration(
                        ProtoDuration.newBuilder()
                            .setSeconds(dedup.inWholeSeconds)
                            .setNanos((dedup.inWholeNanoseconds % 1_000_000_000L).toInt())
                            .build()
                    )
                }
            }
            .build()
}
